### ISSUE TITLE
Extend Unit Tests

### ISSUE THREAD

**Issue by gilligan** [NixOS/nix#3622]
## Nix Unit Testing
Tooling for Unit Testing has been added via #3571 - Unit Tests will help to discover bugs and inconsistencies but also work as documentation and create a reference that makes refactoring easier down the road. This issue is meant to organize the progress on extending unit test coverage on the Nix code base. 

**Note**: Not _all_ code can reasonably be tested or even needs to. Yet a certain coverage is desirable and we are starting from zero.

### How To Test

Nix uses [Google Test](https://github.com/google/googletest). Look at the [Google Test Primer](https://github.com/google/googletest/blob/master/googletest/docs/primer.md) and refer to the existing tests for some guidance on how to write tests.

**Note**: We are writing tests for existing implementations. However you should still avoid looking at the implementation whenever possible as this often leads to tests that (even if unconsciously) re-implement the functions under test. Look instead at the header files and base tests on your assumptions about how the functions should be behaving and encode this in assertions.

### What To Test

The list below just lists the top-level directories from `src/` and is meant as a simple starting point to iterate on.

- [ ] build-remote
- [ ] cpptoml
- [ ] error-demo
- [ ] libexpr
- [ ] libfetchers
- [ ] libmain
- [ ] libstore
- [ ] libutil
  - ~url.hh~  #3603 
  - ~util.hh~ #3571 
  - hash.hh #3620 
  - json.hh #3621 
- [ ] nix
- [ ] nix-build
- [ ] nix-channel
- [ ] nix-collect-garbage
- [ ] nix-copy-closure
- [ ] nix-daemon
- [ ] nix-env
- [ ] nix-instantiate
- [ ] nix-prefetch-url
- [ ] nix-store
- [ ] resolve-system-dependencies

### How to participate

Mostly **just do it**... 

But if you are going to write some more involved tests you might want to mention your intention to do so. If you do write a PR that adds tests, please mention this issue ;)

**Comment by gilligan** [comment]
So as I continue to work on unit tests (and towards covering everything that is reasonably testable in `libutil`) I added a FAQ addressing some topics that came up recently.

I would like to continue to work on tests but I also don't want to do so on my own indefinitely so I would like to move into a direction of recognizing unit tests as mandatory for future contributions/PRs instead of a "nice thing to have".

When the Rust code was added that came with unit tests right away because that's just more common place and easily added to Rust code anyway. Yet now that we have the setup in place there is no reason to not test C++ code.

It would be nice to get some feedback from active maintainers/contributors: @Ericson2314 @matthewbauer @LnL7 @bburdette @zimbatm @edolstra @andir @grahamc

**Comment by edolstra** [comment]
I don't think they should be mandatory. Functional tests are better than unit tests.

**Comment by gilligan** [comment]
> I don't think they should be mandatory. Functional tests are better than unit tests.

I tried my best in the FAQ section above to illustrate the value of and difference between functional and unit tests. Frankly I don't think stating "X is better than Y" without providing _any_ context or explanation is very helpful.

I mean, surely if functional tests were obviously and objectively better than functional tests nobody would waste any time on writing unit tests anymore for any language or project which clearly isn't the case. Because they serve different purposes with different benefits - as i tried to explain above.

**Comment by edolstra** [comment]
Given a finite amount of time, I'd rather spend it writing a functional test that tests what users care about, then testing some internal API, especially if the internal API gets coverage from the functional test too.

It would be different if APIs like libutil were a public API, but they're only used by Nix.

**Comment by gilligan** [comment]
> Given a finite amount of time, I'd rather spend it writing a functional test that tests what users care about, then testing some internal API, especially if the internal API gets coverage from the functional test too.

Based on your argumentation it seems to me like we are writing tests for the sake of increasing code coverage. I think the coverage is just a useful metric.

So **what are we writing tests for?** I think we should be writing different kinds of tests for different purposes (see [The Practical Testing Pyramid](https://martinfowler.com/articles/practical-test-pyramid.html)).

- We should be writing **functional tests** in order to ensure that certain user facing features continue to work.
- We should be writing **unit tests** in order to make sure that single functions/libraries/data structures work in certain expected ways, which by itself also documents how they actually work and how they behave in some edge cases.

I don't think both of those should just be combined into _"something that increases coverage"_. Unit tests can be extremely helpful when refactoring some code. I can just start some watcher to compile/execute the unit tests and get fast and direct feedback on whether or not I refactored correctly or not. I cannot reasonably do this with functional tests. 

> It would be different if APIs like libutil were a public API, but they're only used by Nix.

I guess i would say `libutil` is public within the Nix code base so it's a unit to be tested. Code that is tested is easier to understand, easier to refactor and probably more likely to be correct.

---

**TL;DR**: I want to encourage and establish unit tests for Nix because I think it has the potential to
(1) increase code quality
(2) improve the developer experience
(3) attract more contributors via 1+2

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
